controllers/managedclustermodule_reconciler.go

```diff
@@ -0,0 +1,156 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/manifestwork"
+)
+
+// ManagedClusterModuleReconciler reconciles a ManagedClusterModule object
+type ManagedClusterModuleReconciler struct {
+	client.Client
+
+	manifestAPI manifestwork.ManifestWorkCreator
+}
+
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=work.open-cluster-management.io,resources=manifestworks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch
+
+func NewManagedClusterModuleReconciler(
+	client client.Client,
+	manifestAPI manifestwork.ManifestWorkCreator) *ManagedClusterModuleReconciler {
+	return &ManagedClusterModuleReconciler{
+		Client:      client,
+		manifestAPI: manifestAPI,
+	}
+}
+
+func (r *ManagedClusterModuleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
+
+	logger := log.FromContext(ctx)
+
+	mcm, err := r.getRequestedManagedClusterModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("ManagedClusterModule deleted")
+			return res, nil
+		}
+
+		return res, fmt.Errorf("failed to get the requested CR: %v", err)
+	}
+
+	logger.Info("Requested KMMO ManagedClusterModule")
+
+	clusters, err := r.getSelectedManagedClusters(ctx, mcm)
+	if err != nil {
+		return res, fmt.Errorf("failed to get selected clusters: %v", err)
+	}
+
+	for _, cluster := range clusters.Items {
+		mw := &workv1.ManifestWork{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      mcm.Name,
+				Namespace: cluster.Name,
+			},
+		}
+
+		opRes, err := controllerutil.CreateOrPatch(ctx, r.Client, mw, func() error {
+			return r.manifestAPI.SetManifestWorkAsDesired(ctx, mw, *mcm)
+		})
+
+		if err != nil {
+			return res, fmt.Errorf("failed to create/patch ManifestWork for managed cluster %s: %v", cluster.Name, err)
+		}
+
+		logger.Info("Reconciled ManifestWork", "name", mw.Name, "namespace", mw.Namespace, "result", opRes)
+	}
+
+	if err := r.manifestAPI.GarbageCollect(ctx, *clusters, *mcm); err != nil {
+		return res, fmt.Errorf("failed to garbage collect ManifestWorks with no matching cluster selector: %v", err)
+	}
+
+	return res, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ManagedClusterModuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kmmv1beta1.ManagedClusterModule{}).
+		Owns(&workv1.ManifestWork{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.ManagedCluster{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(r.getSkipUpdatesPredicate())).
```

Comment:
i am probably missing something from the flow: why don't we want to watch for update events on the ManagedClusters?

Reply:
Actually, since we have a cluster labels selector on the `ManagedClusterModule` CR, those labels could change and this would affect the cluster selector. So we should include `ManagedCluster` update events. 

Thanks for the suggestion, I removed this predicate altogether :+1: